stan/math/rev/mat/fun/cov_per.hpp

```diff
@@ -0,0 +1,347 @@
+       * @param l length scale
+       * @param p period
+       */
+      cov_per_vari(const std::vector<T_x>& x,
+                        double sigma,
+                        const T_l& l,
+						const T_p& p)
+        : vari(0.0),
+          size_(x.size()),
+          size_ltri_(size_ * (size_ - 1) / 2),
+          l_d_(value_of(l)), sigma_d_(value_of(sigma)),
+		  p_d_(value_of(p)),
+          sigma_sq_d_(sigma_d_ * sigma_d_),
+          dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  sin_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  cos_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  sin_dist_sq_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+          l_vari_(l.vi_),
+		  p_vari_(p.vi_),
+          cov_lower_(ChainableStack::memalloc_.alloc_array<vari*>(size_ltri_)),
+          cov_diag_(ChainableStack::memalloc_.alloc_array<vari*>(size_)) {
+
+        double neg_two_inv_l_sq = -2.0 / (l_d_ * l_d_);
+        double pi_div_p = M_PI / p_d_;
+
+        size_t pos = 0;
+        for (size_t j = 0; j < size_ - 1; ++j) {
+          for (size_t i = j + 1; i < size_; ++i) {
+          	double dist = distance(x[i], x[j]);
+          	double sin_dist = sin(pi_div_p * dist);
+            double sin_dist_sq = square(sin_dist);
+            dist_[pos] = dist;
+            sin_dist_[pos] = sin_dist;
+          	cos_dist_[pos] = cos(pi_div_p * dist);
+            sin_dist_sq_[pos] = sin_dist_sq;
+            cov_lower_[pos] = new vari(sigma_sq_d_ * std::exp(sin_dist_sq
+                                         * neg_two_inv_l_sq), false);
+            ++pos;
+          }
+        }
+        for (size_t i = 0; i < size_; ++i)
+          cov_diag_[i] = new vari(sigma_sq_d_, false);
+      }
+
+      virtual void chain() {
+        double adjl = 0;
+        double adjp = 0;
+
+        for (size_t i = 0; i < size_ltri_; ++i) {
+          vari* el_low = cov_lower_[i];
+          double prod_add = el_low->adj_ * el_low->val_;
+          adjl += prod_add * sin_dist_sq_[i];
+          adjp += prod_add * sin_dist_[i] * cos_dist_[i] * dist_[i];
+        }
+        double l_d_sq = l_d_ * l_d_;
+        l_vari_->adj_ +=  adjl * 4  / (l_d_sq * l_d_);
+        p_vari_->adj_ +=  adjp * 4 * M_PI / l_d_sq / (p_d_ * p_d_);
+      }
+    };
+
+    /**
+     * Returns a periodic kernel.
+     *
+     * @param x std::vector input that can be used in square distance
+     *    Assumes each element of x is the same size
+     * @param sigma standard deviation
+     * @param l length scale
+     * @param p period
+     * @return squared distance
+     * @throw std::domain_error if sigma <= 0, l <= 0, p <= 0, or
+     *   x is nan or infinite
+     */
+    template <typename T_x>
+    inline typename
+    boost::enable_if_c<boost::is_same<typename scalar_type<T_x>::type,
+                                      double>::value,
+                       Eigen::Matrix<var, -1, -1> >::type
```

Comment:
-1 means Eigen::Dynamic, right? Use the full name

Reply:
done